src/file_analysis/analyzer/x509/functions.bif

```diff
@@ -924,3 +972,125 @@ function x509_set_certificate_cache_hit_callback%(f: string_any_file_hook%) : bo
+## ignoring CNs in the Subject. If no SAN is present, we will use the last CN in the subject
+## for our tests.
+##
+## cert: The X509 certificate opaque handle.
+##
+## hostname: Hostname to check
+##
+## Returns: empty string if the hostname does not match; matched name (which can contain wildcards)
+##          if it did.
+##
+## .. zeek:see:: x509_check_hostname
+function x509_check_cert_hostname%(cert_opaque: opaque of x509, hostname: string%): string
+	%{
+	auto* cert_handle = (zeek::file_analysis::detail::X509Val *) cert_opaque;
+	std::string_view hostview = hostname->ToStdStringView();
+
+	X509* cert = cert_handle->GetCertificate();
+	if ( ! cert )
+		{
+		zeek::emit_builtin_error(zeek::util::fmt("No certificate in opaque"));
+		return zeek::make_intrusive<zeek::StringVal>("");
+		}
+
+	// According to RFC5280 (4.2.1.6) and RFC2818 (3.1), if the SAN is present, the subject
+	// of the certificate is ignored. Let's start by looking at the SAN.
+	auto sanpos = X509_get_ext_by_NID(cert, NID_subject_alt_name, -1);
+	if ( sanpos > -1 )
+		{
+		auto* ex = X509_get_ext(cert, sanpos);
+		if ( ! ex )
+			{
+			zeek::emit_builtin_error(zeek::util::fmt("Could not get SAN from cert"));
+			return zeek::make_intrusive<zeek::StringVal>("");
+			}
+
+		auto *altname = reinterpret_cast<GENERAL_NAMES*>(X509V3_EXT_d2i(ex));
+		if ( ! altname )
+			{
+			zeek::emit_builtin_error(zeek::util::fmt("Could not get names from SAN ext"));
+			return zeek::make_intrusive<zeek::StringVal>("");
+			}
+
+		auto num_names = sk_GENERAL_NAME_num(altname);
+		for ( int i = 0; i < num_names; i++ )
+			{
+			auto *gen = sk_GENERAL_NAME_value(altname, i);
+			assert(gen);
+
+			if ( gen->type != GEN_DNS )
+				continue;
+
+			if ( ASN1_STRING_type(gen->d.ia5) != V_ASN1_IA5STRING )
+				continue;
+
+			std::size_t len = ASN1_STRING_length(gen->d.ia5);
+#if ( OPENSSL_VERSION_NUMBER < 0x10100000L ) || defined(LIBRESSL_VERSION_NUMBER)
+			auto* name = reinterpret_cast<const char*>(ASN1_STRING_data(gen->d.ia5));
+#else
+			auto* name = reinterpret_cast<const char*>(ASN1_STRING_get0_data(gen->d.ia5));
+#endif
+			std::string_view nameview {name, len};
+			if ( check_hostname(hostview, nameview) )
+				return zeek::make_intrusive<zeek::StringVal>(len, name);
+			}
+		}
+	else
+		{
+		// ok, we have to get the last CN from the Subject. Let's do that.
+		auto* subject = X509_get_subject_name(cert);
+		if ( ! subject )
+			{
+			zeek::emit_builtin_error(zeek::util::fmt("Could not get certificate subject"));
+			return zeek::make_intrusive<zeek::StringVal>("");
+			}
+
+		int lastpos = -1;
+		int found_nid = -1;
+		while ( ( lastpos = X509_NAME_get_index_by_NID(subject, NID_commonName, lastpos) ) >= 0 )
+			found_nid = lastpos;
+
+		// found CN
+		if ( found_nid >= 0 )
+			{
+			char buf[2048];
+			BIO *bio = BIO_new(BIO_s_mem());
+			ASN1_STRING_print(bio, X509_NAME_ENTRY_get_data(X509_NAME_get_entry(subject, found_nid)));
+			size_t len = BIO_gets(bio, buf, sizeof(buf));
```

Comment:
Can this fail?

Reply:
BIO_new can in principle fail - and that is the only call that matters. I think it only can fail when we are out of memory - and we have other places in our codebase where we use it without checking the return code.

Reply:
probably still wouldn't hurt to add an `if` around the return type for `BIO_new`, if only to indicate that somebody thought about it.

Reply:
Will do.